File: java/src/test/java/io/binac/leetcode/AddTwoNumbersTest.java
package io.binac.leetcode;

import io.binac.leetcode.util.LinkedLists;
import io.binac.leetcode.util.ListNode;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class AddTwoNumbersTest {
    private static final AddTwoNumbers.Solution1 SOLUTION1 = new AddTwoNumbers.Solution1();

    private final ListNode L11 = LinkedLists.asLinkedList(2, 4, 3);
    private final ListNode L12 = LinkedLists.asLinkedList(5, 6, 4);
    private final ListNode EXPECTED1 = LinkedLists.asLinkedList(7, 0, 8);

    private final ListNode L21 = LinkedLists.asLinkedList(0);
    private final ListNode L22 = LinkedLists.asLinkedList(0);
    private final ListNode EXPECTED2 = LinkedLists.asLinkedList(0);

    private final ListNode L31 = LinkedLists.asLinkedList(9, 9, 9, 9, 9, 9, 9);
    private final ListNode L32 = LinkedLists.asLinkedList(9, 9, 9, 9);
    private final ListNode EXPECTED3 = LinkedLists.asLinkedList(8, 9, 9, 9, 0, 0, 0, 1);

    @Test
    void test1() {
        assertTrue(LinkedLists.equals(EXPECTED1, SOLUTION1.addTwoNumbers(L11, L12)));
        assertTrue(LinkedLists.equals(EXPECTED2, SOLUTION1.addTwoNumbers(L21, L22)));
        assertTrue(LinkedLists.equals(EXPECTED3, SOLUTION1.addTwoNumbers(L31, L32)));
    }
}

File: cpp/tests/AddTwoNumbersTest.cpp
#include "AddTwoNumbers.h"
#include "gtest/gtest.h"

using namespace lcpp;

TEST(AddTwoNumbers, Solution2_1) {
  auto S2_1 = Solution2_1();

  const auto L11 = newLinkedList({2, 4, 3});
  const auto L12 = newLinkedList({5, 6, 4});
  const auto Expected1 = newLinkedList({7, 0, 8});
  const auto Actual1 = S2_1.addTwoNumbers(L11, L12);
  EXPECT_EQ(*Expected1, *Actual1);
  deleteLinkedList(L11);
  deleteLinkedList(L12);
  deleteLinkedList(Expected1);
  deleteLinkedList(Actual1);

  const auto L21 = newLinkedList({0});
  const auto L22 = newLinkedList({0});
  const auto Expected2 = newLinkedList({0});
  const auto Actual2 = S2_1.addTwoNumbers(L21, L22);
  EXPECT_EQ(*Expected2, *Actual2);
  deleteLinkedList(L21);
  deleteLinkedList(L22);
  deleteLinkedList(Expected2);
  deleteLinkedList(Actual2);

  const auto L31 = newLinkedList({9, 9, 9, 9, 9, 9, 9});
  const auto L32 = newLinkedList({9, 9, 9, 9});
  const auto Expected3 = newLinkedList({8, 9, 9, 9, 0, 0, 0, 1});
  const auto Actual3 = S2_1.addTwoNumbers(L31, L32);
  EXPECT_EQ(*Expected3, *Actual3);
  deleteLinkedList(L31);
  deleteLinkedList(L32);
  deleteLinkedList(Expected3);
  deleteLinkedList(Actual3);
}